src/main/java/fi/aalto/cs/apluscourses/utils/PluginAutoInstaller.java

```diff
@@ -0,0 +1,151 @@
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.slf4j.Logger;
+
+public class PluginAutoInstaller {
+
+  private static final Logger logger = APlusLogger.logger;
+
+  private static boolean shouldDownloadPlugin(@NotNull PluginId id) {
+    return !PluginManagerCore.isPluginInstalled(id);
+  }
+
+  private static boolean shouldEnablePlugin(@NotNull PluginId id) {
+    final var plugin = PluginManagerCore.getPlugin(id);
+    return plugin != null && !plugin.isEnabled();
+  }
+
+  /**
+   * Verifies that all plugins with given plugin IDs are installed and enabled. If not, the function
+   * asks the user for consent to install and enable all missing plugins.
+   * @param project The project which initiated the operation.
+   * @param notifier A notifier instance, used to show a notification if something goes wrong.
+   * @param pluginNames List of plugin ID strings. The plugin ID can be found on the plugin's
+   *                    JetBrains Marketplace website.
+   * @param callback An interface implementing functions invoked by this method. The functions
+   *                 are related to showing the consent dialog box to the user.
+   * @return True, if all dependencies are satisfied and there's nothing to be done. False, if
+   * some dependencies were missing, and they were installed or enabled. Null, if the operation
+   * has been cancelled due to missing consent or an error.
+   */
+  @RequiresEdt
+  public static Boolean ensureDependenciesInstalled(@Nullable Project project,
+                                                    @Nullable Notifier notifier,
+                                                    @NotNull List<PluginDependency> pluginNames,
+                                                    @NotNull PluginInstallerCallback callback) {
+    final var pluginIDs = pluginNames.stream().map(p -> PluginId.getId(p.getId())).collect(Collectors.toList());
+
+    // Downloading and enabling plugins are different operations, so we handle these separately.
+    final var pluginsToDownload = pluginIDs.stream()
+        .filter(PluginAutoInstaller::shouldDownloadPlugin).collect(Collectors.toList());
+    final var pluginsToEnable = pluginIDs.stream()
+        .filter(PluginAutoInstaller::shouldEnablePlugin).collect(Collectors.toList());
+
+    pluginsToDownload.forEach(id -> logger.info("Plugin to download: {}", id.getIdString()));
+    pluginsToEnable.forEach(id -> logger.info("Plugin to enable: {}", id.getIdString()));
+
+    if (pluginsToDownload.isEmpty() && pluginsToEnable.isEmpty()) {
+      return true;
+    }
+
+    // allPluginNames = list of PluginDependencies for all plugins in pluginsToDownload or pluginsToEnable
+    final var missingPluginsStream = Stream.concat(pluginsToDownload.stream(), pluginsToEnable.stream());
+    final var allPluginNames = pluginNames.stream()
+        .filter(p -> missingPluginsStream.anyMatch(id -> id.getIdString().equals(p.getId())))
```

Comment:
Is this a performance or correctness issue (i.e. it's incorrect to consume a stream multiple times)? I wrote this code specifically to avoid creating a new collection, but if that's wrong then I beg for forgiveness 🥺
Nevertheless, I'll change it.